src/state.rs

```diff
@@ -68,6 +103,14 @@ pub struct StateSnapshot {
     pub db: salsa::Snapshot<AnalysisDatabase>,
     pub open_files: Snapshot<HashSet<Url>>,
     pub config: Snapshot<Config>,
+    /// Beacon to signal when the snapshot is no longer used
+    pub beacon: Beacon,
+}
+
+impl StateSnapshot {
+    pub(crate) fn signal_finish(&mut self) {
+        self.beacon.signal();
+    }
```

Comment:
Looking at the usage of `Beacon` is it worth introducing at all? You could do `pub callback`, set the callback (or private callback and have a setter) and then do `execute_callback` imo

Reply:
Discussed offline